Upgrade whole webgpu stack

sagudev opened the pull request:
Firstly, this tried to be just a small wgpu bump (0.6 -> 0.16), but so much has changed in webgpu in those two years that this became fully fledged webgpu bump, which is tried to be achieved with minimal changes to the current codebase (no new features only fixing existing, empty stubs where spec require new stuff). Future work:
- using safe callbacks
- fix more wpt tests
- more performant rendering
- implement missing stuff (GPUQueue, GPUSupportedFeatures, ...)

The goal of this PR is to make examples from wgpu working (well those that are also working on Firefox):
- [x] https://mdn.github.io/dom-examples/webgpu-compute-demo/
- [x] https://mdn.github.io/dom-examples/webgpu-render-demo/ (works partially: only blue background; same is in FF so it's probably wgpu problem)
- [x] [Conway's Game of Life](https://sagudev.github.io/briefcase/first-webgpu/9-endgame.html) fixed for naga from https://codelabs.developers.google.com/your-first-webgpu-app
- [ ] ~~https://wgpu.rs/examples-gpu/?example=cube~~ fails due to missing `GPUSupportedFeatures`
- [x] https://wgpu.rs/examples-gpu/?example=hello-compute
- [ ] ~~https://wgpu.rs/examples-gpu/?example=hello-triangle~~ fails due to unimplemented `recreate_contexts`
- [ ] ~~https://wgpu.rs/examples-gpu/?example=mipmap~~ fails due to missing `GPUSupportedFeatures`
- [ ] ~~https://wgpu.rs/examples-gpu/?example=shadow~~ fails due to missing `GPUSupportedFeatures`

to test run with `./mach run -r -- --pref dom.webgpu.enabled` `<URL>`

~~Depends on #29791 which depends on #29770.~~ DONE

We are currently not testing webgpu cts as part of main workflow.

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [ ] These changes fix #___ (GitHub issue number if applicable)

<!-- Either: -->
- [x] There are tests for these changes OR
- [ ] These changes do not require tests because ___

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

sagudev committed:
Upgrade wgpu to 0.16 and refresh whole webgpu implementation

sagudev committed:
Allow noidl files in script/dom/webidls

bors-servo commented:
:umbrella: The latest upstream changes (presumably #29805) made this pull request unmergeable. Please resolve the merge conflicts.

sagudev commented:
```console
Ran 1684 tests finished in 7194.4 seconds.
  • 1657 ran as expected. 0 tests skipped.
  • 22 tests crashed unexpectedly
  • 2 tests unexpectedly okay
  • 5 tests had unexpected subtest results
```

sagudev commented:
Again full of ```
 FAIL [expected NOTRUN] :dimension="2d";format="astc-10x6-unorm-srgb";size=[10,6,257]
  │   │   → assert_unreached: 
  │   │   →   - EXCEPTION: adapter.features is undefined
  │   │   →     supportsFeature@https://web-platform.test:8443/_webgpu/webgpu/webgpu/util/device_pool.js:231:9
  │   │   →     create@https://web-platform.test:8443/_webgpu/webgpu/webgpu/util/device_pool.js:261:10
  │   │   →     async*getOrCreate@https://web-platform.test:8443/_webgpu/webgpu/webgpu/util/device_pool.js:154:34
  │   │   →     acquire@https://web-platform.test:8443/_webgpu/webgpu/webgpu/util/device_pool.js:41:39
  │   │   →     selectDeviceOrSkipTestCase@https://web-platform.test:8443/_webgpu/webgpu/webgpu/gpu_test.js:87:32
  │   │   →     @https://web-platform.test:8443/_webgpu/webgpu/webgpu/api/validation/createTexture.spec.js:614:7
  │   │   →  Reached unreachable code
```

stshine commented:
@bors-servo try=wpt

bors-servo commented:
:lock: Merge conflict

sagudev commented:
@stshine Here is my last try build (does not contain last commit) if you are interested in results: https://github.com/sagudev/servo/actions/runs/5145881293

bors-servo commented:
:umbrella: The latest upstream changes (presumably #29853) made this pull request unmergeable. Please resolve the merge conflicts.

sagudev committed:
Update WebGPU test expectations

sagudev commented:
https://sagudev.github.io/dom-examples/webgpu-compute-demo/ is working in both FF and Servo! Will continue working on canvas at the end of next week.

jdm commented:
That's fantastic!

bors-servo commented:
:umbrella: The latest upstream changes (presumably #29931) made this pull request unmergeable. Please resolve the merge conflicts.

bors-servo commented:
:umbrella: The latest upstream changes (presumably cc585b74d385145e4c3d1af6e92e6b146f354481) made this pull request unmergeable. Please resolve the merge conflicts.

delan commented:
This looks very promising, thanks for working on it!

Are there any parts you would like feedback on yet? How can we help?

> https://sagudev.github.io/dom-examples/webgpu-compute-demo/ is working in both FF and Servo! Will continue working on canvas at the end of next week.

The compute demo works for me too, and works on Firefox if I follow [these steps](https://www.reddit.com/r/wgpu/comments/zjxstc) with a [version containing “a”](https://searchfox.org/mozilla-central/rev/778df10b835ea87def2bfdbb83810ee36bc22c94/build/moz.configure/init.configure#1111-1161) (unlike [firefox-devedition-bin in nixpkgs](https://github.com/NixOS/nixpkgs/blob/51e7e7a385d5ff46f6215d93cbd7d290a00d1645/pkgs/applications/networking/browsers/firefox-bin/devedition_sources.nix)). In both cases, I get no adapters until I tell wgpu where to find libvulkan.so.1:

![image](https://github.com/servo/servo/assets/465303/96b49462-6790-4b27-b9a0-0c522a32c737)

sagudev commented:
> The compute demo works for me too, and works on Firefox if I follow [these steps](https://www.reddit.com/r/wgpu/comments/zjxstc) with a [version containing “a”](https://searchfox.org/mozilla-central/rev/778df10b835ea87def2bfdbb83810ee36bc22c94/build/moz.configure/init.configure#1111-1161) (unlike [firefox-devedition-bin in nixpkgs](https://github.com/NixOS/nixpkgs/blob/51e7e7a385d5ff46f6215d93cbd7d290a00d1645/pkgs/applications/networking/browsers/firefox-bin/devedition_sources.nix)). In both cases, I get no adapters until I tell wgpu where to find libvulkan.so.1:

Firefox dev edition is Firefox beta with additional functionalities. WebGPU works only in nightly FF (versions containing "a" as in alpha).

Compute example was actually patched to work on FF (and servo) due to missing functionalities in naga. The patch is also on upstream now: https://github.com/mdn/dom-examples/pull/221.

All canvas examples that I linked would not be working as they use TLA that is not working on servo. So I am hosting clone of examples on https://sagudev.github.io/briefcase/.

Currently something is wrong with rendering pipeline as I get no errors but screen does not have content as it should have:
![slika](https://github.com/servo/servo/assets/16504129/308ff843-29a3-44ac-bb94-df6726a6791d)
What is black is actually transparent so I program under servo, in FF there is blue color over canvas in chromium based browsers there is also rainbow triangle in the middle.

> Are there any parts you would like feedback on yet? How can we help?

Currently still I know how to progress further, but I will let you know if I get stuck.
This PR is still not reviewable as it is pretty dirty, but I will cleaned it up after I finished work on canvas and marked it as ready.

sagudev committed:
misc

sagudev committed:
MutNullableDom -> DomRefCell<Option<Dom for GPUTexture

sagudev committed:
Direct use of GPUTextureDescriptor

sagudev committed:
Remove config from GPUCanvasContext

sagudev committed:
misc

sagudev committed:
finally blue color

sagudev commented:
Blue color now works!
![slika](https://github.com/servo/servo/assets/16504129/541db2cd-ad65-4c96-802c-6bff98efc8ac)
Now I only need to check other examples and hopefully pass more wpt test.

NOTE TO SELF: wpu debugging: `RUST_LOG="webgpu,wgpu_hal,wgpu_core=trace" ./mach run -d -- --pref dom.webgpu.enabled https://sagudev.github.io/briefcase/`

sagudev commented:
When trying out examples naga is having my problems with some shaders. Here are some common problems and mitigations:
- [automatic type conversations are not supported](https://github.com/gfx-rs/naga/issues/2127) so you need to rewrite literals into type literals: `1` -> `1u` or `1.0`
- swizzling is not supported `cell.xy` -> `vec2(cell.x, cell.y)`
- subtraction must have spaces `cell.x-1u` -> `cell.x - 1u`

sagudev commented:
Wgpu examples might be problematic as they use wasm so it is harder to debug them, but that just means we need more pure js examples.

sagudev committed:
gpubuffer "handle" error

sagudev committed:
GPU object have non-null label

sagudev committed:
gpu limits and info

sagudev committed:
use buffer_size

sagudev commented:
[hello-triangle](https://wgpu.rs/examples-gpu/?example=hello-triangle) example fails due to unimplemented recreate_contexts while others fail due to usage of wgpu example framework, which uses `GPUSupportedFeatures`, which uses `setlike` that is not available in servo.

This concludes work on this PR. Now I just need to do some cleaning and then we can start reviewing.

sagudev committed:
fix warnings

sagudev committed:
Cleanup

sagudev committed:
device destroy

sagudev committed:
fallback adapter

sagudev committed:
mach update-webgpu write webgpu commit hash in file

sagudev commented:
I think this is now ready for review 🎉. Review as a whole as commits are messy.

Latest try build: https://github.com/sagudev/servo/actions/runs/5849369462

Locally more WPT test PASS, but expectations are currently set for CI which has more FAILS. I will be investigating this in the mean time.

sagudev committed:
Mising deps in CI for webgpu tests

sagudev committed:
Updated expectations

sagudev commented:
> Locally more WPT test PASS, but expectations are currently set for CI which has more FAILS. I will be investigating this in the mean time.

Fixed and updated expectations.

sagudev commented:
Rebased.

jdm commented:
FYI: I am reviewing this.

jdm commented on components/script/dom/webidls/GPU.webidl:
```diff
@@ -5,15 +5,17 @@
 // https://gpuweb.github.io/gpuweb/#gpu-interface
 [Exposed=(Window, DedicatedWorker), Pref="dom.webgpu.enabled"]
 interface GPU {
+    // May reject with DOMException
```
Can we add a Throws annotation, in that case?

sagudev replied:
This was written because ff has throw here when fails getting global. We do not do that, so I removed this.

jdm commented on components/webgpu/lib.rs:
```diff
@@ -371,7 +376,14 @@ impl<'a> WGPU<'a> {
+                factory,
+                InstanceDescriptor {
+                    backends: wgt::Backends::PRIMARY,
+                    dx12_shader_compiler: Dx12Compiler::default(),
```
Will this work on non-Windows backends?

sagudev replied:
Yes, this option is ignored on non-windows platforms.

jdm reviewed (approved):
This was a pretty high level review, because the scale of the changes here was really big and I'm not really familiar with webgpu. I think it's unambiguously better to merge these changes and use an up-to-date wgpu dependency, so thank you for putting the effort in to make that happen!

sagudev commented on components/script/dom/gpuadapter.rs:
```diff
@@ -183,10 +183,10 @@ impl GPUAdapterMethods for GPUAdapter {
                 WebGPURequest::RequestDevice {
                     sender,
                     adapter_id: self.adapter,
+                    label: desc.label.clone(), // TODO(sagudev)
```
I am not sure labels handling is the best one so I plan to rewrite/unify label handling in follow up PR.

sagudev commented on components/script/dom/gpuadapter.rs:
```diff
@@ -68,57 +69,106 @@ impl GPUAdapter {
+                    "maxComputeWorkgroupsPerDimension" => {
+                        desc.limits.max_compute_workgroups_per_dimension = v
+                    },
+                    _ => { /* no-op */ },
```
Hm, that's not in latest version. Commits in between can be messy.

sagudev commented on components/script/dom/gpucanvascontext.rs:
```diff
@@ -122,7 +123,7 @@ pub struct GPUCanvasContext {
     texture_desc: DomRefCell<Option<GPUTextureDescriptor>>,
     /// https://gpuweb.github.io/gpuweb/#dom-gpucanvascontext-currenttexture-slot
-    texture: MutNullableDom<GPUTexture>,
+    texture: DomRefCell<Option<Dom<GPUTexture>>>, // TODO(sagudev): sea of MutNullableDom ???
```
That's again cross commit change in this PR, but idk the reason it just seemed easier to handle. I'm still a little confused with the sea of all Dom types.

sagudev replied:
Actually now that I reviewed it's usage I'm more confident that this is the right choice.

jdm replied:
Using MutNullableDom when possible is often preferred, since there's no way to trigger borrow-related panics.

sagudev replied:
I think we can get away with MutNullableDom.

sagudev commented on components/script/dom/gpuadapter.rs:
```diff
@@ -68,57 +69,106 @@ impl GPUAdapter {
+                GPUFeatureName::Depth_clip_control => {
+                    features.insert(wgt::Features::DEPTH_CLIP_CONTROL)
+                },
+                GPUFeatureName::Depth24unorm_stencil8 => { /* not implemented in wgpu */ },
```
It should.

sagudev commented on components/script/dom/gpucanvascontext.rs:
```diff
@@ -133,38 +214,86 @@ impl LayoutCanvasRenderingContextHelpers for LayoutDom<'_, GPUCanvasContext> {
-                    .lock()
-                    .create_buffer_id(descriptor.device.id().0.backend()),
-            );
+        // Step 2&3
```
This is from old commit but I did cleaned some stuff up.

sagudev committed:
Fixups

sagudev commented:
I created new try build here: https://github.com/sagudev/servo/actions/runs/5918122292

NOTE: Before landing https://github.com/servo/servo/pull/29795/commits/1d0699adca2bf4f741fac61b9f7340e9fd423605 needs to be removed.

jdm commented on components/script/dom/gpuadapter.rs:
```diff
@@ -90,7 +90,11 @@ impl GPUAdapterMethods for GPUAdapter {
+                GPUFeatureName::Depth24unorm_stencil8 => {
+                    promise.reject_error(Error::Type(String::from(
+                        "depth24unorm_stencil8 is not supported by wgpu",
+                    )));
```
I suspect we need to return promise immediate after rejecting it (and later in this function as well) to avoid running more code that may trigger additional rejection.

sagudev committed:
early reject

sagudev committed:
DomRefCell<Option<Dom -> MutNullableDom for GPUTexture

sagudev commented:
New try build running: https://github.com/sagudev/servo/actions/runs/5918319598

sagudev commented:
> NOTE: Before landing [1d0699a](https://github.com/servo/servo/commit/1d0699adca2bf4f741fac61b9f7340e9fd423605) needs to be removed.

This commit is now removed. Ready for landing!
